workflow.go

```diff
@@ -0,0 +1,161 @@
+package subscription
+
+import (
+	"time"
+
+	"go.temporal.io/sdk/workflow"
+)
+
+var (
+	// These thresholds/periods are purposefully short to allow for easier
+	// testing and demonstration of Continue-As-New. The Temporal Events in
+	// this application are small enough that it's unlikely to hit the event
+	// *size* limit, and so in the "real" version of this,
+	// EVENT_HISTORY_THRESHOLD should be much closer to 50k.
+	EVENT_HISTORY_THRESHOLD      = 1000
+	HISTORY_CHECK_PERIOD_SECONDS = 5
+
+	// TODO making this a global feels icky. What's the better pattern, given workflow.Go doesn't accept additional args?
+	CONTINUE_AS_NEW_CHANNEL workflow.Channel
+)
+
+type Messages struct {
+	UpdateInfo UpdateSignal
+}
+
+func (s *Messages) HandleSignals(ctx workflow.Context, canChannel workflow.Channel) {
+	logger := workflow.GetLogger(ctx)
+
+	for {
+		workflow.NewSelector(ctx).
+			AddReceive(workflow.GetSignalChannel(ctx, "update"), func(c workflow.ReceiveChannel, _ bool) {
+				var updateInfo UpdateSignal
+				c.Receive(ctx, &updateInfo)
+				logger.Info("Received update signal", "data", updateInfo)
+				// TODO: apply updates
+			}).
+			AddReceive(workflow.GetSignalChannel(ctx, "cancel"), func(c workflow.ReceiveChannel, _ bool) {
+				var cancel UpdateSignal
+				c.Receive(ctx, &cancel)
+				logger.Info("Received cancel signal", "data", cancel)
+
+				if cancel.CancelSubscription {
+					// TODO: cancel subscription
+				}
+			}).
+			Select(ctx)
+	}
+}
+
+func pollHistorySize(ctx workflow.Context) {
+	logger := workflow.GetLogger(ctx)
+	info := workflow.GetInfo(ctx)
+
+	for {
+		_ = workflow.Sleep(ctx, time.Duration(HISTORY_CHECK_PERIOD_SECONDS*int(time.Second)))
+
+		logger.Info("Current history length", "History Length", info.GetCurrentHistoryLength())
+		if info.GetCurrentHistoryLength() > EVENT_HISTORY_THRESHOLD {
+			CONTINUE_AS_NEW_CHANNEL.Send(ctx, true)
+		}
+	}
+}
+
+func SubscriptionWorkflow(ctx workflow.Context, subscription SubscriptionInfo) error {
+	ao := workflow.ActivityOptions{
+		StartToCloseTimeout: 10 * time.Second,
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	logger := workflow.GetLogger(ctx)
+	logger.Info("Subscription workflow started", "customer", subscription.Customer, "subscription", subscription)
+
+	// Immediately kick off update/cancel handlers, since those can come in at
+	// any time. Channel used to signal that Continue-As-New is needed
+	var m Messages
+	CONTINUE_AS_NEW_CHANNEL = workflow.NewChannel(ctx)
+	workflow.GoNamed(ctx, "signalPoller", func(ctx workflow.Context) {
+		m.HandleSignals(ctx, CONTINUE_AS_NEW_CHANNEL)
+	})
+
+	continueAsNew := false
+	continueHandler := func(c workflow.ReceiveChannel, _ bool) {
+		var canMessage bool
+		c.Receive(ctx, &canMessage)
```

Comment:
You should always handle errors

Reply:
workflow.ReceiveChannel.Receive doesn't return errors though? Moot in this exact instance after my rewrite, but question may apply elsewhere.

Reply:
Sorry, my mistake, ignore me.